### PR TITLE
1664 Implement tracing

### LINKED ISSUE #1664: Implement debug_Trace methods
Debug trace methods are in the code and we need to enable them for the historic state

The goal is to be compatible with debug_Trace methods of geth, so we can be used by Dune Analytics, Tenderly and other tools.

### PR DESCRIPTION
**Pull request by kladkogex**
# Documentation

The documentation for this feature is provided here: 

https://github.com/skalenetwork/skaled/blob/1664_eth_trace_current/docs/tracing.md

# Tests

An automated test that tests tracing has been added to 

test/historicstate/hardhat/scripts/trace.ts

The test deploys a sample smart contract and tests all tracers.

The test runs as part of Github Actions test.yml

### CODE REVIEW

File: libweb3jsonrpc/Debug.cpp [hunk 1]
```diff
@@ -1,262 +1,278 @@
...
-        op.disableStorage = _json["disableStorage"].asBool();
-    if ( !_json["disableMemory"].empty() )
-        op.disableMemory = _json["disableMemory"].asBool();
-    if ( !_json["disableStack"].empty() )
-        op.disableStack = _json["disableStack"].asBool();
-    if ( !_json["fullStorage"].empty() )
-        op.fullStorage = _json["fullStorage"].asBool();
-    return op;
+void Debug::checkPrivilegedAccess() const {
+    if ( m_enablePrivilegedApis ) {
```
Comment (olehnikolaiev) [inline libweb3jsonrpc/Debug.cpp]: shouldn't it be `!m_enablePrivelegedApis`?

File: README.md [hunk 2]
```diff
@@ -60,7 +60,7 @@ If you have already cloned the repo and forgot to pass `--recurse-submodules`, e
 sudo apt update
 sudo apt install autoconf build-essential cmake libprocps-dev libtool texinfo wget yasm flex bison btrfs-progs python3 python3-pip gawk git vim doxygen 
 sudo apt install make build-essential cmake pkg-config libgnutls28-dev libssl-dev unzip zlib1g-dev libgcrypt20-dev docker.io gcc-9 g++-9 gperf clang-format-11 gnutls-dev
-sudo apt install nettle-dev libhiredis-dev redis-server google-perftools libgoogle-perftools-dev lcov
+sudo apt install nettle-dev libhiredis-dev redis-server google-perftools libgoogle-perftools-dev lcov sudo apt-get install libv8-dev
```
Comment (DmytroNazarenko) [inline README.md:63]: Can you please fix typo

File: libethereum/Client.cpp [hunk 3]
```diff
@@ -1285,6 +1292,65 @@ ExecutionResult Client::call( Address const& _from, u256 _value, Address _dest,
...
+        auto er = historicBlock.executeHistoricCall( bc().lastBlockHashes(), _t, _tracer, 0 );
+        return _tracer->getJSONResult();
+    } catch ( ... ) {
+        cwarn << boost::current_exception_diagnostic_information();
+        throw;
+    }
+}
+
+Json::Value Client::traceBlock( BlockNumber _blockNumber, Json::Value const& _jsonTraceConfig ) {
+    Block previousBlock = blockByNumber( _blockNumber - 1 );
```
Comment (DmytroNazarenko) [inline libethereum/Client.cpp:1314]: What if _blockNumber == 0?

File: libethereum/Client.cpp [hunk 4]
```diff
@@ -1285,6 +1292,65 @@ ExecutionResult Client::call( Address const& _from, u256 _value, Address _dest,
...
+
+    // cache results for better peformance
+    string key = to_string( _blockNumber ) + traceOptions.toString();
+
+    auto cachedResult = m_blockTraceCache.getIfExists( key );
+    if ( cachedResult.has_value() ) {
+        return std::any_cast< Json::Value >( cachedResult );
+    }
+
+    for ( unsigned k = 0; k < transactions.size(); k++ ) {
```
Comment (DmytroNazarenko) [inline libethereum/Client.cpp:1332]: Do we need to handle 0-gas transactions?

File: libhistoric/AlethStandardTrace.cpp [hunk 5]
```diff
@@ -1,95 +1,411 @@
...
+
+vector< uint8_t > AlethStandardTrace::extractSmartContractMemoryByteArrayFromStackPointer(
+    const LegacyVM* _vm ) {
+    STATE_CHECK( _vm )
+
+    vector< uint8_t > result{};
+
+    if ( _vm->stackSize() > 2 ) {
+        auto b = ( uint32_t ) _vm->getStackElement( 0 );
+        auto s = ( uint32_t ) _vm->getStackElement( 1 );
```
Comment (DmytroNazarenko) [inline libhistoric/AlethStandardTrace.cpp:139]: Please, use more descriptive names for b & s

File: libhistoric/AlethStandardTrace.cpp [hunk 6]
```diff
@@ -1,95 +1,411 @@
...
+    Json::Value result = Json::Value( Json::ValueType::objectValue );
+
+    for ( auto&& entry : m_tracePrinters ) {
+        TracePrinter& printer = entry.second;
+        printer.print( result, _er, _statePre, _statePost );
+        m_jsonTrace[printer.getJsonName()] = result;
+        result.clear();
+    }
+}
+const h256& AlethStandardTrace::getTxHash() const {
```
Comment (DmytroNazarenko) [inline libhistoric/AlethStandardTrace.cpp:388]: Please add blank lines between functions for readability

File: libhistoric/FunctionCallRecord.cpp [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+
+// this class collect information about functions called during execution
+
+namespace dev::eth {
+
+void FunctionCallRecord::setGasUsed( uint64_t _gasUsed ) {
+    m_gasUsed = _gasUsed;
+}
+uint64_t FunctionCallRecord::getFunctionGasLimit() const {
```
Comment (DmytroNazarenko) [inline libhistoric/FunctionCallRecord.cpp:38]: Please add blank lines between functions

**Review by DmytroNazarenko — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (DmytroNazarenko) [inline README.md:63]: Please, fix the typo

File: libhistoric/AlethStandardTrace.cpp [hunk 8]
```diff
@@ -1,95 +1,421 @@
...
+
+vector< uint8_t > AlethStandardTrace::extractSmartContractMemoryByteArrayFromStackPointer(
+    const LegacyVM* _vm ) {
+    STATE_CHECK( _vm )
+
+    vector< uint8_t > result{};
+
+    if ( _vm->stackSize() > 2 ) {
+        auto b = ( uint32_t ) _vm->getStackElement( 0 );
+        auto s = ( uint32_t ) _vm->getStackElement( 1 );
```
Comment (DmytroNazarenko) [inline libhistoric/AlethStandardTrace.cpp:139]: Please, use more descriptive names for b & s

**Review by DmytroNazarenko — CHANGES_REQUESTED** [review]

**Pull request merged**